File: python/src/wallet_observer/policy.py
"""Balance floors and logical recommendations; no execution capability."""

from decimal import Decimal

from ._validation import ObserverError, calculate, integer, unique
from .observer import build_snapshot
from .types import (
    AnalysisReport,
    Finding,
    MovementRecommendation,
    PortfolioPolicy,
    PortfolioSnapshot,
)


def analyze(
    snapshot: PortfolioSnapshot,
    policy: PortfolioPolicy,
    analyzed_at_ms: int,
) -> AnalysisReport:
    if snapshot.schema_version != 1 or policy.schema_version != 1:
        raise ObserverError("schema_version")
    integer(analyzed_at_ms)
    integer(policy.maximum_snapshot_age_ms, 1)
    # Validate persisted inputs before using their balances for policy analysis.
    rebuilt = build_snapshot(snapshot.observed_at_ms, list(snapshot.balances))
    if (
        rebuilt.wallet_nav != snapshot.wallet_nav
        or rebuilt.total_valued_nav_usd != snapshot.total_valued_nav_usd
    ):
        raise ObserverError("inconsistent_snapshot_nav")
    age = max(0, analyzed_at_ms - snapshot.observed_at_ms)
    findings = []
    if age > policy.maximum_snapshot_age_ms:
        findings.append(
            Finding(
                "critical",
                "stale_snapshot",
                None,
                None,
                f"snapshot age {age} ms exceeds {policy.maximum_snapshot_age_ms} ms",
            )
        )
    holdings = {(b.wallet_id, b.asset_id): b.quantity for b in snapshot.balances}
    minimums = {}
    for requirement in policy.minimum_balances:
        unique([requirement.wallet_id])
        unique([requirement.asset_id])
        key = (requirement.wallet_id, requirement.asset_id)
        if key in minimums:
            raise ObserverError("duplicate_minimum")
        minimums[key] = requirement.minimum_quantity
        observed = holdings.get(key)
        if observed is None:
            findings.append(
                Finding(
                    "critical",
                    "missing_balance",
                    *key,
                    "required wallet/asset balance is absent",
                )
            )
        elif observed < requirement.minimum_quantity:
            findings.append(
                Finding(
                    "warning",
                    "below_minimum",
                    *key,
                    f"observed {observed} is below minimum {requirement.minimum_quantity}",
                )
            )
    working = dict(holdings)
    recommendations = []
    zero = Decimal(0)
    for (receiver, asset), minimum in sorted(minimums.items()):
        deficit = calculate(minimum, "-", working.get((receiver, asset), zero))
        if not deficit:
            continue
        donors = sorted(w for w, a in working if a == asset and w != receiver)
        for donor in donors:
            available = calculate(working[(donor, asset)], "-", minimums.get((donor, asset), zero))
            if not available:
                continue
            quantity = min(deficit, available)
            recommendations.append(MovementRecommendation(donor, receiver, asset, quantity))
            working[(donor, asset)] = calculate(working[(donor, asset)], "-", quantity)
            working[(receiver, asset)] = calculate(
                working.get((receiver, asset), zero), "+", quantity
            )
            deficit = calculate(deficit, "-", quantity)
            if not deficit:
                break
    return AnalysisReport(
        1, analyzed_at_ms, age, not findings, tuple(findings), tuple(recommendations)
    )

File: src/policy.rs
use crate::types::{
    AnalysisReport, Finding, MovementRecommendation, PortfolioPolicy, PortfolioSnapshot, Severity,
};
use rust_decimal::Decimal;
use std::collections::{BTreeMap, HashMap};
use thiserror::Error;

#[derive(Debug, Error, PartialEq, Eq)]
pub enum AnalysisError {
    #[error("snapshot and policy schema_version must be 1")]
    SchemaVersion,
    #[error("maximum_snapshot_age_ms must be positive")]
    InvalidMaximumAge,
    #[error("duplicate balance for wallet={wallet_id}, asset={asset_id}")]
    DuplicateBalance { wallet_id: String, asset_id: String },
    #[error("minimum quantity must not be negative")]
    NegativeMinimum,
    #[error("duplicate minimum for wallet={wallet_id}, asset={asset_id}")]
    DuplicateMinimum { wallet_id: String, asset_id: String },
}

pub fn analyze(
    snapshot: &PortfolioSnapshot,
    policy: &PortfolioPolicy,
    analyzed_at_ms: i64,
) -> Result<AnalysisReport, AnalysisError> {
    if snapshot.schema_version != 1 || policy.schema_version != 1 {
        return Err(AnalysisError::SchemaVersion);
    }
    if policy.maximum_snapshot_age_ms <= 0 {
        return Err(AnalysisError::InvalidMaximumAge);
    }
    let snapshot_age_ms = analyzed_at_ms
        .saturating_sub(snapshot.observed_at_ms)
        .max(0);
    let mut findings = Vec::new();
    if snapshot_age_ms > policy.maximum_snapshot_age_ms {
        findings.push(Finding {
            severity: Severity::Critical,
            code: "stale_snapshot".to_string(),
            wallet_id: None,
            asset_id: None,
            detail: format!(
                "snapshot age {} ms exceeds {} ms",
                snapshot_age_ms, policy.maximum_snapshot_age_ms
            ),
        });
    }

    let mut holdings = HashMap::new();
    for balance in &snapshot.balances {
        let key = (balance.wallet_id.clone(), balance.asset_id.clone());
        if holdings.insert(key.clone(), balance.quantity).is_some() {
            return Err(AnalysisError::DuplicateBalance {
                wallet_id: key.0,
                asset_id: key.1,
            });
        }
    }
    let mut minimums = HashMap::new();
    for requirement in &policy.minimum_balances {
        if requirement.minimum_quantity.is_sign_negative() {
            return Err(AnalysisError::NegativeMinimum);
        }
        let key = (requirement.wallet_id.clone(), requirement.asset_id.clone());
        if minimums
            .insert(key.clone(), requirement.minimum_quantity)
            .is_some()
        {
            return Err(AnalysisError::DuplicateMinimum {
                wallet_id: key.0,
                asset_id: key.1,
            });
        }
        let observed = holdings
            .get(&(requirement.wallet_id.clone(), requirement.asset_id.clone()))
            .copied();
        match observed {
            None => findings.push(Finding {
                severity: Severity::Critical,
                code: "missing_balance".to_string(),
                wallet_id: Some(requirement.wallet_id.clone()),
                asset_id: Some(requirement.asset_id.clone()),
                detail: "required wallet/asset balance is absent".to_string(),
            }),
            Some(quantity) if quantity < requirement.minimum_quantity => {
                findings.push(Finding {
                    severity: Severity::Warning,
                    code: "below_minimum".to_string(),
                    wallet_id: Some(requirement.wallet_id.clone()),
                    asset_id: Some(requirement.asset_id.clone()),
                    detail: format!(
                        "observed {} is below minimum {}",
                        quantity, requirement.minimum_quantity
                    ),
                });
            }
            _ => {}
        }
    }

    let recommendations = recommendations(&holdings, &minimums);
    Ok(AnalysisReport {
        schema_version: 1,
        analyzed_at_ms,
        snapshot_age_ms,
        healthy: findings.is_empty(),
        findings,
        recommendations,
    })
}

fn recommendations(
    holdings: &HashMap<(String, String), Decimal>,
    minimums: &HashMap<(String, String), Decimal>,
) -> Vec<MovementRecommendation> {
    let mut working = holdings.clone();
    let mut requirements: Vec<_> = minimums.iter().collect();
    requirements.sort_by(|left, right| left.0.cmp(right.0));
    let mut output = Vec::new();
    for ((receiver, asset), minimum) in requirements {
        let current = working
            .get(&(receiver.clone(), asset.clone()))
            .copied()
            .unwrap_or_default();
        let mut deficit = (*minimum - current).max(Decimal::ZERO);
        if deficit.is_zero() {
            continue;
        }
        let mut donors = BTreeMap::new();
        for ((wallet, holding_asset), quantity) in &working {
            if wallet == receiver || holding_asset != asset {
                continue;
            }
            let donor_minimum = minimums
                .get(&(wallet.clone(), asset.clone()))
                .copied()
                .unwrap_or_default();
            let available = (*quantity - donor_minimum).max(Decimal::ZERO);
            if !available.is_zero() {
                donors.insert(wallet.clone(), available);
            }
        }
        for (donor, available) in donors {
            let quantity = deficit.min(available);
            output.push(MovementRecommendation {
                from_wallet_id: donor.clone(),
                to_wallet_id: receiver.clone(),
                asset_id: asset.clone(),
                quantity,
                reason: "restore configured minimum balance".to_string(),
                executable: false,
                requires_external_approval: true,
            });
            *working.entry((donor, asset.clone())).or_default() -= quantity;
            *working
                .entry((receiver.clone(), asset.clone()))
                .or_default() += quantity;
            deficit -= quantity;
            if deficit.is_zero() {
                break;
            }
        }
    }
    output
}
